src/Npgsql/NpgsqlConnector.cs

```diff
@@ -369,45 +373,77 @@ internal async Task Open(NpgsqlTimeout timeout, bool async, CancellationToken ca
+                            _backendSecretKey = keyDataMsg.BackendSecretKey;
+                            msg = await ReadMessage(async);
+                        }
+                        if (msg.Code != BackendMessageCode.ReadyForQuery)
+                            throw new NpgsqlException($"Received backend message {msg.Code} while expecting ReadyForQuery. Please file a bug.");
 
-                State = ConnectorState.Ready;
+                        State = ConnectorState.Ready;
 
-                await LoadDatabaseInfo(timeout, async);
+                        await LoadDatabaseInfo(timeout, async);
+                        UpdateServerPrimaryStatus();
 
-                if (Settings.Pooling && DatabaseInfo.SupportsDiscard)
-                    GenerateResetMessage();
-                Counters.NumberOfNonPooledConnections.Increment();
-                Counters.HardConnectsPerSecond.Increment();
-                Log.Trace($"Opened connection to {Host}:{Port}");
+                        if (IsAppropriateFor(Settings.TargetServerType) == false)
+                        {
+                            SoftCleanup();
+                            throw new NpgsqlException($"Host {ConnectedHost} was not of compatible type. Got: {ConnectedServerType} Expected: {Settings.TargetServerType}");
+                        }
 
-                // If an exception occurs during open, Break() below shouldn't close the connection, which would also
-                // update pool state. Instead we let the exception propagate and get handled by the calling pool code.
-                // We use an extra state flag because the connector's State varies during the type loading query
-                // above (Executing, Fetching...). Note also that Break() gets called from ReadMessageLong().
-                _isConnecting = false;
+                        if (Settings.Pooling && DatabaseInfo.SupportsDiscard)
+                            GenerateResetMessage();
+                        Counters.NumberOfNonPooledConnections.Increment();
+                        Counters.HardConnectsPerSecond.Increment();
+                        Log.Trace($"Opened connection to {ConnectedHost}:{Port}");
+
+                        // If an exception occurs during open, Break() below shouldn't close the connection, which would also
+                        // update pool state. Instead we let the exception propagate and get handled by the calling pool code.
+                        // We use an extra state flag because the connector's State varies during the type loading query
+                        // above (Executing, Fetching...). Note also that Break() gets called from ReadMessageLong().
+                        _isConnecting = false;
+                        break;
+                    }
+                    catch (SocketException)
+                    {
+                        if (hosts.Last() == host) 
```

Comment:
This would lead to unexpected behaviour if a host is listed multiple times. That might be an unlikely scenario, but possible: e.g. maybe host A is nearby and hosts B,C,D are far away, so I set my hosts to "A,B,C,D,A" in order to try A again just before giving up.

Reply:
Hmm that _is_ true, out of curiosity is that an actual scenario that you could see yourself using?

Reply:
We'd have to test it first, but maybe!